File: lab04.1/ciphers_csharp/VigenereCipher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ciphers_csharp {
    internal static class VigenereCipher {
        private const string CYRILLIC_ALPHABET = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ";
        private const string LATIN_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

        private static string _getKey(string message, string keyword) {
            string key = "";
            while (key.Length < message.Length) {
                key += keyword;
            }
            return key.Substring(0, message.Length);
        }

        private static string _vigenereCipher(string oldMessage, string keyword, int key, string alphabet) {
            string fullKey = _getKey(oldMessage, keyword);
            string newMessage = "";
            for (int i = 0; i < oldMessage.Length; i++) {
                char msgCh = oldMessage[i];
                char keyCh = fullKey[i];
                string msgAlphabet = char.IsUpper(msgCh) ? alphabet.ToUpper() : alphabet.ToLower();
                string keyAlphabet = char.IsUpper(keyCh) ? alphabet.ToUpper(): alphabet.ToLower();
                int msgChIndex = msgAlphabet.IndexOf(msgCh);
                int keyChIndex = keyAlphabet.IndexOf(keyCh);
                char newCh;
                if (msgChIndex == -1) {
                    newCh = msgCh;
                }
                else {
                    int newChIndex = (msgChIndex + key * keyChIndex) % alphabet.Length;
                    if (newChIndex < 0) {
                        newChIndex = alphabet.Length + newChIndex;
                    }
                    newCh = msgAlphabet[newChIndex];
                }
                newMessage += newCh;
            }
            return newMessage;
        }

        public static string VigenereCipherEncryption(string decryptedMessage, string keyword, int key, string alphabet) {
            return _vigenereCipher(
                decryptedMessage,
                keyword,
                key,
                alphabet
            );
        }

        public static string VigenereCipherDecryption(string encryptedMessage, string keyword, int key, string alphabet) {
            return _vigenereCipher(
                encryptedMessage,
                keyword,
                -key,
                alphabet
            );
        }

        public static string VigenereCipherCyrillicEncryption(string decryptedMessage, string keyword, int key) {
            return VigenereCipherEncryption(
                decryptedMessage,
                keyword,
                key,
                CYRILLIC_ALPHABET
            );
        }

        public static string VigenereCipherCyrillicDecryption(string encryptedMessage, string keyword, int key) {
            return VigenereCipherDecryption(
                encryptedMessage,
                keyword,
                key,
                CYRILLIC_ALPHABET
            );
        }

        public static string VigenereCipherLatinEncryption(string decryptedMessage, string keyword, int key) {
            return VigenereCipherEncryption(
                decryptedMessage,
                keyword,
                key,
                LATIN_ALPHABET
            );
        }

        public static string VigenereCipherLatinDecryption(string encryptedMessage, string keyword, int key) {
            return VigenereCipherDecryption(
                encryptedMessage,
                keyword,
                key,
                LATIN_ALPHABET
            );
        }
    }

}

File: lab04.1/ciphers_python/vigenere_cipher.py
CYRILLIC_ALPHABET = "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"
LATIN_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"


def _get_key(message: str, keyword: str) -> str:
    key = ""
    while len(key) < len(message):
        key += keyword
    return key[:len(message)]


def _vigenere_cipher(*, old_message: str, keyword: str, key: int, alphabet: str) -> str:
    full_key = _get_key(old_message, keyword)
    new_message = ""
    for (msg_ch, key_ch) in zip(old_message, full_key):
        msg_alphabet = alphabet.upper() if msg_ch.isupper() else alphabet.lower()
        key_alphabet = alphabet.upper() if key_ch.isupper() else alphabet.lower()
        msg_ch_index = msg_alphabet.find(msg_ch)
        key_ch_index = key_alphabet.find(key_ch)
        if msg_ch_index == -1:
            new_ch = msg_ch
        else:
            new_ch_index = (msg_ch_index + key * key_ch_index) % len(alphabet)
            new_ch = msg_alphabet[new_ch_index]
        new_message += new_ch
    return new_message


def vigenere_cipher_encryption(*, decrypted_message: str, keyword: str, key: int, alphabet: str) -> str:
    return _vigenere_cipher(
        old_message=decrypted_message,
        keyword=keyword,
        key=key,
        alphabet=alphabet,
    )


def vigenere_cipher_decryption(*, encrypted_message: str, keyword: str, key: int, alphabet: str) -> str:
    return _vigenere_cipher(
        old_message=encrypted_message,
        keyword=keyword,
        key=-key,
        alphabet=alphabet,
    )


def vigenere_cipher_cyrillic_encryption(decrypted_message: str, keyword: str, key: int) -> str:
    return vigenere_cipher_encryption(
        decrypted_message=decrypted_message,
        keyword=keyword,
        key=key,
        alphabet=CYRILLIC_ALPHABET,
    )


def vigenere_cipher_cyrillic_decryption(encrypted_message: str, keyword: str, key: int) -> str:
    return vigenere_cipher_decryption(
        encrypted_message=encrypted_message,
        keyword=keyword,
        key=key,
        alphabet=CYRILLIC_ALPHABET,
    )


def vigenere_cipher_latin_encryption(decrypted_message: str, keyword: str, key: int) -> str:
    return vigenere_cipher_encryption(
        decrypted_message=decrypted_message,
        keyword=keyword,
        key=key,
        alphabet=LATIN_ALPHABET,
    )


def vigenere_cipher_latin_decryption(encrypted_message: str, keyword: str, key: int) -> str:
    return vigenere_cipher_decryption(
        encrypted_message=encrypted_message,
        keyword=keyword,
        key=key,
        alphabet=LATIN_ALPHABET,
    )
